content/blog/2022/2022-08-03-feature-flags-in-nodejs-react-client-with-launchdarkly.md

```diff
@@ -0,0 +1,305 @@
+
+```bash
+npm start
+```
+
+The development server will start up at http://localhost:3000 and we can see the home page.
+
+{{% image alt="React UI" src="images/posts/nodejs-frontend-launchdarkly/React_Logo.png" %}}
+
+Next we will add the [launchdarkly-js-client-sdk](https://github.com/launchdarkly/js-client-sdk) package to use client-side methods to fetch the feature flag variations:
+
+```bash
+npm install launchdarkly-js-client-sdk
+```
+
+
+
+## String-based Feature Flag to Hide/Display Content on the Fly
+
+In our first use-case, let’s try to simply define a string feature flag in LaunchDarkly UI. We will simply list some users in the React UI. If any of the users matches the user name defined in the feature flag variation, then we will encrypt that user name in the UI. It would be pretty simple but enough to understand how we can enable/disable a feature in web UI.
+
+The feature flag defined in LaunchDarkly UI would look something like this:
+
+{{% image alt="String Feature Flag LaunchDarkly" src="images/posts/nodejs-frontend-launchdarkly/Hide_User.png" %}}
+
+By default, flags are only available to server-side SDKs. When we create a flag, we can choose to expose the flag to SDKs which use client-side IDs, SDKs which use mobile keys, or both. If we are using a client-side or mobile SDK, we must expose our feature flags in order for the client-side or mobile SDKs to evaluate them. As per the [documentation](https://docs.launchdarkly.com/home/getting-started/feature-flags#making-flags-available-to-client-side-and-mobile-sdks), in case of React UI or any Javascript UI, we need to enable *“SDKs using Client-side ID”* option. So in this case, we will fetch the *Client-Side ID* from LaunchDarkly Projects page and define it as part of the code. We should also make sure that the targeting option is enabled for the flag.
+
+{{% image alt="LaunchDarkly Keys" src="images/posts/nodejs-frontend-launchdarkly/LaunchDarkly_Keys.png" %}}
+
+Next we will update our `App.js` to fetch the feature flag variations from LaunchDarkly and display the list of users. 
+
+```javascript
+import React, { Component } from 'react';
+import './App.css';
+import * as LDClient from 'launchdarkly-js-client-sdk';
+
+class App extends Component {
+  constructor() {
+    super()
+    this.state = {
+      selectedSortOrder: null,
+      users: [
+        { name: 'John Doe', added: new Date('2022-7-27') },
+        { name: 'Allen Witt', added: new Date('2022-6-30') },
+        { name: 'Cheryl Strong', added: new Date('2022-7-02') },
+        { name: 'Marty Byrde', added: new Date('2022-5-03') },
+        { name: 'Wendy Byrde', added: new Date('2022-6-03') },
+      ]
+    }
+  }
+  componentDidMount() {
+    const user = {
+      // UI based user
+      key: 'user_a'
+    }
+    // SDK requires Client-side ID for UI call
+    this.ldclient = LDClient.initialize('62e9289ade464c10d842c2b3', user);
```

Comment:
You can argue that you left the key in this example because it's a client-side key and that is safe to share (after having explained the differences between client and server keys above)

Reply:
I have encrypted the key as we did for previous article.